oak/server/rust/oak_runtime/src/node/grpc/client.rs

```diff
@@ -0,0 +1,258 @@
+        address: Uri,
+        domain_name: String,
+        root_tls_certificate: Certificate,
+        initial_reader: Handle,
+    ) -> Self {
+        Self {
+            config_name: config_name.to_string(),
+            runtime: runtime.clone(),
+            address,
+            domain_name,
+            root_tls_certificate,
+            receiver: Receiver::new(runtime, initial_reader),
+        }
+    }
+
+    /// Main Node worker thread.
+    fn worker_thread(mut self) {
+        let pretty_name = pretty_name_for_thread(&thread::current());
+
+        // Create an Async runtime for executing futures.
+        // https://docs.rs/tokio/
+        let mut async_runtime = tokio::runtime::Builder::new()
+            // Use simple scheduler that runs all tasks on the current-thread.
+            // https://docs.rs/tokio/0.2.16/tokio/runtime/index.html#basic-scheduler
+            .basic_scheduler()
+            // Enables the I/O driver.
+            // Necessary for using net, process, signal, and I/O types on the Tokio runtime.
+            .enable_io()
+            .build()
+            .expect("Couldn't create an Async runtime");
+
+        // Listen to incoming gRPC invocations.
+        info!(
+            "{}: Starting a gRPC client pseudo-Node on: {}",
+            pretty_name, self.address
+        );
+        let result = async_runtime.block_on(self.handle_loop());
+        info!(
+            "{}: Exiting gRPC client pseudo-Node thread {:?}",
+            pretty_name, result
+        );
+
+        self.runtime.exit_node();
+    }
+
+    /// Main loop that handles gRPC invocations.
+    async fn handle_loop(&mut self) -> Result<(), OakStatus> {
+        // Connect to an external gRPC service.
+        let mut handler = loop {
+            if let Ok(handler) = self.connect().await {
+                break handler;
+            }
+            // Retry to connect after 5 seconds.
+            thread::sleep(Duration::from_secs(5));
+        };
+
+        loop {
+            // Read a gRPC invocation from the [`GrpcClientNode::receiver`].
+            let invocation = self.read_invocation()?;
+
+            // Receive a request from the invocation channel.
+            let request = invocation.receive_request()?;
+
+            // Send an unary request to an external gRPC service and wait for the response.
+            let response = handler.unary_request(request).await?;
+
+            // Send a response back to the invocation channel.
+            invocation.send_response(response)?;
+        }
+    }
+
+    // Creates a TLS connection to an external gRPC service.
+    async fn connect(&mut self) -> Result<GrpcRequestHandler, OakStatus> {
+        debug!("Connecting to {} [{}]", self.domain_name, self.address);
+
+        // Create a TLS configuration.
+        let tls_config = ClientTlsConfig::new()
+            .domain_name(self.domain_name.to_string())
+            .ca_certificate(self.root_tls_certificate.clone());
```

Comment:
The CA certificate is also only needed if it the CA is not in the set of already trusted CAs of the host OS. This should also probably support optional logic.

Reply:
In order to use OS certificates I can enable the following feature:
```
feature = "tls-roots"
```
https://github.com/hyperium/tonic/blob/04a8c0c82a4007f48c3bf3539a3f2312746fedd1/tonic/src/transport/service/tls.rs#L59-L62

But I thought that we are specifically trying to ignore OS certs.
cc @tiziano88